controllers/containerizedworkload_controller.go

```diff
@@ -0,0 +1,156 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"time"
+
+	cpv1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/go-logr/logr"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	oamv1alpha2 "github.com/oam-dev/core-resource-controller/api/v1alpha2"
+)
+
+const (
+	oamReconcileWait = 30 * time.Second
+)
+
+// Reconcile error strings.
+const (
+	errRenderWorkload  = "cannot render workload"
+	errUpdateStatus    = "cannot apply status"
+	errApplyDeployment = "cannot apply the deployment"
+	errApplyService    = "cannot apply the service"
+	errGCDeployment    = "cannot clean up stale deployments"
+	errScaleDeployment = "cannot scale the deployment"
+)
+
+// ContainerizedWorkloadReconciler reconciles a ContainerizedWorkload object
+type ContainerizedWorkloadReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core.oam.dev,resources=containerizedworkloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core.oam.dev,resources=containerizedworkloads/status,verbs=get;update;patch
+func (r *ContainerizedWorkloadReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
```

Comment:
We typically set a timeout or deadline on this context to ensure the reconcile eventually finishes.

Reply:
Makes sense.  For some reason, many of the k8s library functions don't take a context as the first parameter....

Reply:
> We typically set a timeout or deadline on this context to ensure the reconcile eventually finishes.

Sounds good but AFAIK, the context deadline needs to be handled explicitly somewhere. Does the controller-runtime frame work it?

Reply:
You know, I had always assumed it was given that they take `context.Context` as an argument. Looks like they actually only added support for the unstructured client (which handles kinds that aren't registered with the scheme) a few days ago in https://github.com/kubernetes-sigs/controller-runtime/pull/812. In this case we're using the typed client, which has had support for a while now per https://github.com/kubernetes-sigs/controller-runtime/issues/40.

Reply:
yeah,  even the client-go doesn't take context until recently (https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20200123-client-go-ctx.md)